src/lib/ptree/trace.lua

```diff
@@ -0,0 +1,84 @@
+-- Use of this source code is governed by the Apache 2.0 license; see COPYING.
+
+module(...,package.seeall)
+
+local lib = require('core.lib')
+local json = require("lib.ptree.json")
+
+local Trace = {}
+local trace_config_spec = {
+   file = {required=true},
+   file_mode = {default="w"},
+}
+
+function new (conf)
+   local conf = lib.parse(conf, trace_config_spec)
+   local ret = setmetatable({}, {__index=Trace})
+   ret.id = 0
+   ret.output = io.open(conf.file, conf.file_mode)
+   return ret
+end
+
+local function listen_directive_for_rpc(rpc_id, args)
+   local ret = { path=args.path, schema=args.schema, revision=args.revision }
+   if rpc_id == 'get-config' then
+      ret.verb = 'get'
+      return ret
+   elseif rpc_id == 'set-config' then
+      ret.verb, ret.value = 'set', args.config
+      return ret
+   elseif rpc_id == 'add-config' then
+      ret.verb, ret.value = 'add', args.config
+      return ret
+   elseif rpc_id == 'remove-config' then
+      ret.verb = 'remove'
+      return ret
+   elseif rpc_id == 'get-state' then
+      ret.verb = 'get-state'
+      return ret
+   else
+      return nil
+   end
+end
+
+function Trace:record(id, args)
+   assert(self.output, "trace closed")
+   local obj = listen_directive_for_rpc(id, args)
+   if not obj then return end
+   obj.id = tostring(self.id)
+   self.id = self.id + 1
+   json.write_json_object(self.output, obj)
+   self.output:write('\n')
+   self.output:flush()
+end
+
+function Trace:close()
+   self.output:close()
+   self.output = nil
+end
+
+function selftest ()
+   print('selftest: lib.ptree.trace')
+   local S = require('syscall')
+
+   local tmp = os.tmpname()
+   local trace = new({file=tmp})
+   trace:record("foo", {bar="baz"})
+   trace:record("qux", {bar="baz", zog="100"})
```

Comment:
The test fails cause `trace:record` calls `listen_directive_for_rpc` which only admits a limited number of verbs and args. Is either necessary to change the test arguments or broad `listen_directive_for_rpc` to admit more verbs and args, dunno.

Reply:
Fixed test.